presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java

```diff
@@ -0,0 +1,215 @@
+    {
+        Expression expressionWithNotPushedDown = ExpressionTreeRewriter.rewriteWith(new PushDownNegationsExpressionRewriter(), expression);
+        return ExpressionTreeRewriter.rewriteWith(new ExtractCommonPredicatesExpressionRewriter(), expressionWithNotPushedDown);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = context.rewrite(node.getSource());
+            Expression simplified = simplifyPredicates(node.getPredicate());
+            return new FilterNode(node.getId(), source, simplified);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = context.rewrite(node.getSource());
+            Map<Symbol, Expression> assignments = node.getAssignments().entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, entry -> simplifyPredicates(entry.getValue())));
+            return new ProjectNode(node.getId(), source, assignments);
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
+        {
+            Expression originalConstraint = null;
+            if (node.getOriginalConstraint() != null) {
+                originalConstraint = simplifyPredicates(node.getOriginalConstraint());
+            }
+            return new TableScanNode(
+                    node.getId(),
+                    node.getTable(),
+                    node.getOutputSymbols(),
+                    node.getAssignments(),
+                    node.getLayout(),
+                    node.getCurrentConstraint(),
+                    originalConstraint);
+        }
+    }
+
+    private static class PushDownNegationsExpressionRewriter
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteNotExpression(NotExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getValue() instanceof LogicalBinaryExpression) {
+                LogicalBinaryExpression child = (LogicalBinaryExpression) node.getValue();
+                List<Expression> predicates = extractPredicates(child);
+                List<Expression> negatedPredicates = predicates.stream()
+                        .map(p -> treeRewriter.rewrite((Expression) new NotExpression(p), context))
+                        .collect(toList());
+                return combinePredicates(child.getType().flip(), negatedPredicates);
+            }
+            else if (node.getValue() instanceof ComparisonExpression) {
+                ComparisonExpression child = (ComparisonExpression) node.getValue();
+                return new ComparisonExpression(
+                        child.getType().negate(),
```

Comment:
i think you need to custom handle the case where the type is IS_DISTINCT_FROM. The negation of that type needs to handle the NULL value properly.